src/editor/codemirror/copypaste.ts

```diff
@@ -0,0 +1,78 @@
+import { Extension } from "@codemirror/state";
+import { EditorView } from "@codemirror/view";
+import { lineNumFromUint8Array } from "../../common/text-util";
+import { deployment } from "../../deployment";
+import { CodeInsertType } from "./dnd";
+import { calculateChanges } from "./edits";
+
+interface PasteContext {
+  code: string;
+  type: CodeInsertType;
+}
+
+const getCodeFromHtml = (
+  innerHTML: string | undefined
+): PasteContext | undefined => {
+  if (!innerHTML) {
+    return;
+  }
+  const parent: HTMLElement = document.createElement("html");
+  parent.innerHTML = innerHTML;
+  return {
+    code: (parent as HTMLElement).querySelector("code")?.textContent || "",
+    type:
+      ((parent as HTMLElement)
+        .querySelector("code")
+        ?.getAttribute("data-type") as CodeInsertType) || "unknown",
+  };
+};
+
+const copyPasteHandlers = () => {
+  const textEncoder = new TextEncoder();
+  return [
+    EditorView.domEventHandlers({
+      paste(event, view) {
+        if (
+          !view.state.facet(EditorView.editable) ||
+          event.clipboardData?.getData("text")
+        ) {
+          return;
+        }
+        event.preventDefault();
```

Comment:
I feel like we shouldn't call this unless we have text to insert. Otherwise we might prevent some CM logic. Probably not, given the earlier tests, but I think structuring like that makes sense. So, "is it the weird special case" -> e.preventDefault() and do the action. Otherwise we just return and CM runs.

Reply:
This change is fine given the plan on https://github.com/microbit-foundation/python-editor-next/pull/766#discussion_r898917406.